src/circuit_proof/r1cs.rs

```diff
@@ -182,79 +215,73 @@ impl ConstraintSystem {
     }
 
     // This function can only be called once per ConstraintSystem instance.
-    pub fn create_proof_input<R: Rng + CryptoRng>(
+    pub fn prove<R: Rng + CryptoRng>(
         mut self,
-        pedersen_gens: &PedersenGenerators,
+        v_blinding: &Vec<Scalar>,
+        gen: &Generators,
+        transcript: &mut Transcript,
         rng: &mut R,
-    ) -> (
-        Circuit,
-        Result<ProverInput, R1CSError>,
-        Result<VerifierInput, R1CSError>,
-    ) {
-        // If `n`, the number of multiplications, is not 0 or 2, then pad the circuit.
-        let n = self.aL_assignments.len();
-        if !(n == 0 || n.is_power_of_two()) {
-            let pad = n.next_power_of_two() - n;
-            let zer = Scalar::zero();
-            for _ in 0..pad {
-                self.assign_multiplier(Ok(zer), Ok(zer), Ok(zer));
-            }
+    ) -> Result<(CircuitProof, VerifierInput), R1CSError> {
+        if v_blinding.len() != self.commitments_count() {
+            return Err(R1CSError::IncorrectInputSize);
         }
-        let v_blinding: Vec<Scalar> = (0..self.commitments_count())
-            .map(|_| Scalar::random(rng))
-            .collect();
-
+        // create circuit params
         let circuit = self.create_circuit();
-        let prover_input = self.create_prover_input(&v_blinding);
-        let verifier_input = self.create_verifier_input(pedersen_gens, &v_blinding);
+        let prover_input = self.create_prover_input(&v_blinding)?;
+        let verifier_input = self.create_verifier_input(&gen.pedersen_gens, &v_blinding)?;
 
-        (circuit, prover_input, verifier_input)
+        // TODO: use error handling instead of unwrap
+        let circuit_proof =
+            CircuitProof::prove(&gen, transcript, rng, &circuit, &prover_input).unwrap();
```

Comment:
Maybe add a `map_err` right away since the outer function already returns Result and avoid the `unwrap`?